pkg/policy/snapshots.go

```diff
@@ -0,0 +1,236 @@
+package policy
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/jeremywohl/flatten"
+)
+
+func (ce *Executor) persistSnapshot(ctx context.Context, path string, table string) error {
+	ef, err := os.OpenFile(filepath.Join("%s/", path, fmt.Sprintf("table_%s.csv", table)), os.O_CREATE|os.O_WRONLY, 0777)
+	if err != nil {
+		ce.log.Error("error opening file.", "error", err)
+		return err
+	}
+	defer ef.Close()
+	query := fmt.Sprintf("COPY (select * from %s) TO STDOUT DELIMITER '|' CSV HEADER", table)
+	ce.log.Debug("exporting", "table", table, "query", query)
+
+	err = ce.storeOutput(ctx, ef, query)
+	if err != nil {
+		return fmt.Errorf("error exporting file: %w", err)
+	}
+	return nil
+
+}
+func (ce *Executor) StoreSnapshot(ctx context.Context, path string, tables []string) error {
+	if len(tables) == 0 {
+		return errors.New("no tables to snapshot")
+	}
+
+	for _, table := range tables {
+
+		err := ce.persistSnapshot(ctx, path, table)
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func (ce *Executor) RestoreSnapshot(ctx context.Context, source string) error {
+
+	ef, err := os.OpenFile(source, os.O_RDONLY, 0777)
+	if err != nil {
+		ce.log.Error("error opening file.", "error", err)
+		return err
+	}
+	defer ef.Close()
+	fileName := path.Base(source)
+	if !strings.HasPrefix(fileName, "table_") || !strings.HasSuffix(fileName, ".csv") {
+		ce.log.Error("truncating", "source", source, "file", fileName)
+		return fmt.Errorf("invalid filename: %q", fileName)
+	}
+	source_name := strings.TrimPrefix(strings.TrimSuffix(fileName, ".csv"), "table_")
+	truncQuery := fmt.Sprintf("TRUNCATE %s CASCADE", source_name)
+	ce.log.Debug("truncating", "table", source_name, "query", truncQuery)
+	err = ce.conn.Exec(ctx, truncQuery)
+	if err != nil {
+		return fmt.Errorf("error importing file %q: %w", fileName, err)
+	}
+
+	query := fmt.Sprintf("copy \"%s\" from stdin DELIMITER '|' CSV HEADER;", source_name)
+	ce.log.Debug("importing", "table", source_name, "query", query)
+
+	err = ce.conn.RawCopyFrom(ctx, ef, query)
+	if err != nil {
+		return fmt.Errorf("error importing file: %w", err)
+	}
+
+	return nil
+}
+
+func cleanQuery(query string) string {
+	// This removes comments from the query
+	// single line queries are not supported in pg_dump
+	var re = regexp.MustCompile(`(?s)\/\*.*?\*\/|--.*?\n`)
+	query = re.ReplaceAllString(query, "")
+	query = strings.TrimSuffix(query, ";")
+	query = strings.TrimSpace(query)
+	query = strings.TrimSuffix(query, ";")
+
+	return strings.TrimSpace(query)
+}
+
+func (ce *Executor) ExtractTableNames(ctx context.Context, query string) ([]string, error) {
```

Comment:
would add another nit, this function is very long, would still prefer the pg_depend method using the SQL as its cleaner and has less chances for issues. Moreover it will require less calls to the DB if we are working with views.